sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/expressions/expressions.scala

```diff
@@ -0,0 +1,194 @@
+ */
+private[sql] abstract class SingleColumnTransform(ref: NamedReference) extends Transform {
+
+  def reference: NamedReference = ref
+
+  override lazy val references: Array[NamedReference] = Array(ref)
+
+  override lazy val arguments: Array[Expression] = Array(ref)
+
+  override lazy val describe: String = name + "(" + reference.describe + ")"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class BucketTransform(
+    numBuckets: Literal[Int],
+    columns: Seq[NamedReference]) extends Transform {
+
+  override lazy val name: String = "bucket"
+
+  override lazy val references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override lazy val arguments: Array[Expression] = numBuckets +: columns.toArray
+
+  override lazy val describe: String = s"bucket(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class ApplyTransform(
+    name: String,
+    args: Seq[Expression]) extends Transform {
+
+  override lazy val arguments: Array[Expression] = args.toArray
+
+  override lazy val references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override lazy val describe: String = s"$name(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class IdentityTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "identity"
+  override lazy val describe: String = ref.describe
+}
+
+private[sql] final case class YearTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "year"
+}
+
+private[sql] final case class MonthTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "month"
+}
+
+private[sql] final case class DateTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "date"
+}
+
+private[sql] final case class DateHourTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "date_hour"
+}
+
+private[sql] final case class LiteralValue[T](value: T, dataType: DataType) extends Literal[T] {
+  override def describe: String = {
+    if (dataType.isInstanceOf[StringType]) {
+      s"'$value'"
+    } else {
+      s"$value"
+    }
+  }
+  override def toString: String = describe
+}
+
+private[sql] final case class FieldReference(fieldNames: Seq[String]) extends NamedReference {
+  override def describe: String = fieldNames.map(quote).mkString(".")
+  override def toString: String = describe
+
+  private def quote(part: String): String = {
+    if (part.contains(".")) {
+      s"`$part`"
```

Comment:
what if the name part contains ` as well?

Reply:
nvm, this is same as `UnresolvedAttribute.name`

Reply:
Actually, I think you're right. This is intended to be a representation that we could send through the parser to return the exact same value, so this should double to escape it. That is done in `IdentifierWithDatabase` for other identifiers.